File: libs/libppf/src/parser.cc
/*
    Programmable Property Files
    Copyright (C) 2004 Marcus Andersson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "ppf/parser.h"
#include "ppf/bundle.h"
#include "ppf/ppfexception.h"

#include <fstream>

using namespace ppf;

Parser::Parser()
{
  interp = Tcl_CreateInterp();
  if (Tcl_Init(interp) == TCL_ERROR) {
    throw PPFException("Failed to init TCL interpreter.", 3);
  }
  rootBundle = new Bundle(interp);  
}

BoundBundle Parser::parse(string filename)
{
  string line;
  ifstream file(filename.c_str());
  Bundle* currentBundle;
  
  if (file.fail()) {
    throw
      PPFException(string("Failed to open file: ") + filename,
				    5);
  }

  while (!file.eof()) {
    getline(file, line);
    line = trim(line);
    if (line.length() > 0 && line[0] != '#') {
      StringList bundles;
      string property;
      string expression;
      parse(line, &bundles, &property, &expression);
      currentBundle = rootBundle;
      for (StringList::iterator name = bundles.begin();
	   name != bundles.end(); name++) {
	currentBundle = currentBundle->newBundle(*name);
      }
      currentBundle->newProperty(property, expression);
    }
  }
  file.close();
  return BoundBundle(rootBundle, 0, "");
}

Parser::~Parser()
{
  delete rootBundle;
}

string Parser::trim(string line)
{
  while (line.length() > 0 && (line[0] == ' ' || line[0] == '\t')) {
    line = line.substr(1);
  }
  return line.substr(0, line.length() - 1);
}

void Parser::parse(string line,
		   StringList* bundles, string* property, string* expression)
{
  string token;
  while (line.find(" ") < line.length()) {
    token = line.substr(0, line.find(" "));
    replace(&token, "[", "\\[");
    replace(&token, "]", "\\]");
    line = line.substr(line.find(" ") + 1);
    if (token == "=") {
      break;
    }
    bundles->push_back(token);
  }
  if (token != "=") {
    throw
      PPFException(string("Missing '=' sign: ") + line, 2);
  }
  *expression = line;
  *property = bundles->back();
  bundles->pop_back();
}

void Parser::replace(string* token, string match, string replacement)
{
  string::size_type pos = token->find(match);
  while(pos < token->length()) {
    *token = token->replace(pos, match.length(), replacement);
    pos = token->find(match, pos + replacement.length());
  }
}

File: nomad/src/org/nomad/patch/format/parser/Parser.java
/* Copyright (C) 2006 Christian Schneider
 * 
 * This file is part of Nomad.
 * 
 * Nomad is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Nomad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Nomad; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
 * Created on Feb 28, 2006
 */
package org.nomad.patch.format.parser;

public class Parser {

	/**
	 * starting from pos, pos is incremented while any whitespace character is found.
	 * the position of the next non whitespace character is returned.
	 * 
	 * @param str
	 * @param pos position in str (pos<len)
	 * @param len length of str
	 * @return position of next non whitespace character. pos might be greater or equal than len 
	 */
	public final static int ignoreWhitespace(String str, int pos, int len)
	{
		while (pos<len)
		{
			if (Character.isWhitespace(str.charAt( pos ))) 
			{
				pos ++;
			} 
			else break ;

		}
		return pos ;
	}

	// same as above, but reads from pos to 0 (right to left), may return -1
	public final static int ignoreWhitespaceRewind(String str, int pos)
	{
		while (pos>=0)
		{
			if (Character.isWhitespace(str.charAt( pos ))) 
			{
				pos --;
			} 
			else break ;

		}
		return pos ;
	}

	/**
	 * reads digits from left to right and returns the integer number
	 * 
	 * @param str
	 * @param record
	 * @return -1 if no digit was found
	 */
	public final static int parseDigits(String str, Record record)
	{
		int number = 0;
		int spos = record.spos ;
		char c;
		while (spos < record.slen) 
		{
			c = str.charAt( spos );
			if (Character.isDigit(c))
			{
				number = (number*10)+(c-'0');
				spos ++;
			} 
			else break ;
		}
		
		if (record.spos != spos) 
		{
			record.spos = spos ;
			return number ;
		} 
		else return -1 ;
	}
	
	public final static int parseDigitsRewind(String str, Record record)
	{
		int number = 0;
		int spos = record.spos ;
		int shift = 1;
		char c;
		while (spos > 0) 
		{
			c = str.charAt( spos );
			if (Character.isDigit(c))
			{
				number = number + ((c-'0') * shift);
				shift ++ ;
				spos --;
			} 
			else break ;
		}
		
		if (record.spos != spos) 
		{
			record.spos = spos ;
			return number ;
		} 
		else return -1 ;
	}

	public final static boolean parseCharacterRewind(String str, char c, Record record)
	{
		if (record.spos < 0)
			return false;
		
		if (str.charAt(record.spos)!=c)
			return false;
		
		record.spos --;
		return true ;
	}
	
}
